src/theory/quantifiers/sygus/cegis_unif.cpp

```diff
@@ -0,0 +1,293 @@
+  {
+    unsigned j = enum_consider[i];
+    Node e = enums[j];
+    Node v = enum_values[j];
+    std::vector<Node> enum_lems;
+    d_sygus_unif.notifyEnumeration(e, v, enum_lems);
+    /* the lemmas must be guarded by the active guard of the enumerator */
+    Assert(d_enum_to_active_guard.find(e) != d_enum_to_active_guard.end());
+    Node g = d_enum_to_active_guard[e];
+    for (unsigned j = 0, size = enum_lems.size(); j < size; ++j)
+    {
+      enum_lems[j] = nm->mkNode(OR, g.negate(), enum_lems[j]);
+    }
+    lems.insert(lems.end(), enum_lems.begin(), enum_lems.end());
+  }
+  /* build candidate solution */
+  Assert(candidates.size() == 1);
+  Node vc = d_sygus_unif.constructSolution();
+  if (vc.isNull())
+  {
+    return false;
+  }
+  candidate_values.push_back(vc);
+  return true;
+}
+
+Node CegisUnif::purifyFuncApp(
+    Node n,
+    std::vector<Node>& model_guards,
+    std::unordered_map<Node, Node, NodeHashFunction>& cache)
+{
+  Trace("cegis-unif-purify") << "PurifyFuncApp : " << n << "\n";
+  std::unordered_map<Node, Node, NodeHashFunction>::const_iterator it =
+      cache.find(n);
+  if (it != cache.end())
+  {
+    Trace("cegis-unif-purify") << "... already visited " << n << "\n";
+    return it->second;
+  }
+  unsigned i, size = n.getNumChildren();
+  Kind k = n.getKind();
+  bool childChanged = false;
+  std::vector<Node> children;
+  /* Recurse */
+  for (i = 0; i < size; ++i)
+  {
+    Node child = purifyFuncApp(n[i], model_guards, cache);
+    children.push_back(child);
+    childChanged = childChanged || child != n[i];
+  }
+  Node nb;
+  if (childChanged)
+  {
+    if (n.hasOperator())
+    {
+      children.insert(children.begin(), n.getOperator());
+    }
+    nb = NodeManager::currentNM()->mkNode(k, children);
+    Trace("cegis-unif-purify") << "PurifyFuncApp : transformed " << n
+                               << " into " << nb << "\n";
+  }
+  else
+  {
+    nb = n;
+  }
+  /* get model value if d_candidate applied over constants */
+  if (k == APPLY_UF && size > 0 && nb[0] == d_candidate)
+  {
+    Trace("cegis-unif-purify") << "PurifyFuncApp : candidate app " << nb
+                               << "\n";
+    for (i = 1; i < size; ++i)
+    {
+      if (!nb[i].isConst())
+      {
+        break;
+      }
+    }
+    if (i == size)
+    {
+      Node nv = d_parent->getModelValue(nb);
+      Trace("cegis-unif-purify") << "PurifyFuncApp : model value for " << nb
+                                 << " is " << nv << "\n";
+      /* test if different, then update model_guards */
+      if (nv != nb)
+      {
+        Trace("cegis-unif-purify") << "PurifyFuncApp : adding model eq\n";
+        model_guards.push_back(
+            NodeManager::currentNM()->mkNode(EQUAL, nv, nb).negate());
+        nb = nv;
+      }
+    }
+  }
+  /* put in cache */
+  Trace("cegis-unif-purify") << "... caching [" << n << "] = " << nb << "\n";
+  cache[n] = nb;
+  return nb;
+}
+
+Node CegisUnif::purifyLemma(
```

Comment:
This function is a bit hard to follow. 
Can you merge this function and the one above and add a flag "ensureConst" that ensures that the return value is a constant?